app/routes/login/callback.tsx

```diff
@@ -0,0 +1,79 @@
+import type { ActionArgs } from '@remix-run/node';
+import { createUserSession } from '~/session.server';
+import { parseLoginResponse } from '~/saml.server';
+import { redirect } from '@remix-run/node';
+import { createUser, getUserByUsername } from '~/models/user.server';
+
+/* This is the post route that the SAML response is bound to. It comes back as formData.
+  We attempt to extract the SAML response into a json format that we can then use:
+
+The response from SimpleSAMLPhp
+{
+  conditions: {
+    notBefore: '2023-02-16T17:09:02Z',
+    notOnOrAfter: '2023-02-16T17:14:32Z'
+  },
+  response: {
+    id: '_cb60d5bbb95d8c820bbb306640ea535e9ffd6f23ca',
+    issueInstant: '2023-02-16T17:09:32Z',
+    destination: 'http://localhost:8080/login/callback',
+    inResponseTo: '_4be46ad1-4c2c-42c7-ba1f-a23c31161fcd'
+  },
+  audience: 'http://host.docker.internal:8080/sp',
+  issuer: 'http://localhost:8081/simplesaml/saml2/idp/metadata.php',
+  nameID: 'user1@myseneca.ca',
+  sessionIndex: {
+    authnInstant: '2023-02-16T17:09:32Z',
+    sessionNotOnOrAfter: '2023-02-17T01:09:32Z',
+    sessionIndex: '_6befc9246d7bd30b9e0faea72d0836cddf3ff8d556'
+  },
+  attributes: {
+    uid: '1',
+    eduPersonAffiliation: 'group1',
+    'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress': 'user1@myseneca.ca',
+    email: 'user1@myseneca.ca',
+    'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname': 'Johannes',
+    'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/surname': 'Kepler',
+    'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name': 'user1@myseneca.ca',
+    sAMAccountName: 'user1',
+    'http://schemas.microsoft.com/identity/claims/displayname': 'Johannes Kepler'
+  }
+}
+*/
+export const action = async ({ request }: ActionArgs) => {
+  if (request.method !== 'POST') {
+    // Request method is not post, why are you here?
+    return redirect('/');
+  }
+
+  const formData = await request.formData();
+  const body = Object.fromEntries(formData);
+  const samlResponse = await parseLoginResponse(body);
+
+  // Try and extract the username and see if there is an existing user by that name
+  if (!samlResponse.attributes.sAMAccountName) {
+    // TODO: Make this redirect to access denied page
+    return redirect('/');
+  }
+  const username = samlResponse.attributes.sAMAccountName;
+  // get or create user
+  let user = await getUserByUsername(username);
+
+  // If not create one
+  if (!user) {
+    user = await createUser(
+      username,
+      samlResponse.attributes['http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname'],
+      samlResponse.attributes['http://schemas.xmlsoap.org/ws/2005/05/identity/claims/surname'],
+      samlResponse.attributes['http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress']
+    );
+  }
+
+  // Either way create a session
+  return createUserSession({
+    request: request,
+    username: username,
+    remember: false,
+    redirectTo: '/',
```

Comment:
Do we have access to the `redirectTo` param here?  When you are on a page, and get redirected to the login, it usually keeps this param so you can get sent back to the place where you were.  It should be in `params` I think, but check my math.

Reply:
I remembered that Chris's example had something for this using RelayState so that the redirectUrl can persist to the callback. I added that.